CentOS/duffy
# Session workflow

This implements a good deal of #65: accepting specifications of requested nodes when creating sessions and returning the associated nodes with sessions from the API. It doesn't implement a controller for updating sessions (to let tenants set them inactive) yet, though.

Additionally, implement filling a DB with test data (as test fixtures and an extension to the `setup-db` command) and some refinement to tests.

# How to Test

- Delete the database (file) and perhaps recreate it (PostgreSQL)
- Setup the database and fill it with test data, e.g.:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> duffy -c etc/duffy-psql-config.yaml setup-db --test-data
Creating database schema
Setting up database migrations
Creating test data
Caution! Created tenants with deterministic API keys:
	admin: ae6c10d0-0b13-554c-b976-a05d8a18f0cc
	tenant: a8b9899d-b128-59a1-aa86-754920b7f5ed
Don't use in production!
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)>
```
- Verify the test data is there, e.g.:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> psql duffy
psql (13.4)
Type "help" for help.

duffy=> select * from nodes;
          created_at           | retired_at | id |    type    |               hostname                
|    ipaddr    | state  | comment | distro_type | distro_version 
-------------------------------+------------+----+------------+---------------------------------------
+--------------+--------+---------+-------------+----------------
 2021-12-21 12:18:22.036233+01 |            |  1 | opennebula | node-opennebula-small-7.example.net   
| 192.168.1.17 | active |         | fedora      | 35
[ ... ]
 2021-12-21 12:18:22.036233+01 |            | 72 | opennebula | node-opennebula-medium-2.example.net  
| 192.168.2.12 | active |         | centos      | 8Stream
(72 rows)

duffy=>
```
- Start the app server
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> duffy -c etc/duffy-psql-config.yaml serve
 * Starting Duffy...
 * Host address : 127.0.0.1
 * Port number  : 8080
 * Log level    : warning
 * Serving API docs on http://127.0.0.1:8080/docs
```
- Request a session with a certain set of nodes:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> http --json post http://localhost:8080/api/v1/sessions tenant_id=2 nodes_specs:='[{"quantity": 1, "type": "physical", "distro_type": "fedora", "distro_version": "35"}, {"quantity": 2, "type": "virtual", "flavour": "medium", "distro_type": "fedora", "distro_version": "35"}]'
HTTP/1.1 201 Created
content-length: 769
content-type: application/json
date: Tue, 21 Dec 2021 12:22:12 GMT
server: uvicorn

{
    "action": "post",
    "session": {
        "active": true,
        "created_at": "2021-12-21T11:22:13.362065+00:00",
        "id": 1,
        "nodes": [
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "hostname": "node-seamicro-2.example.net",
                "ipaddr": "192.168.0.12",
                "type": "seamicro"
            },
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "flavour": "medium",
                "hostname": "node-opennebula-medium-3.example.net",
                "ipaddr": "192.168.2.13",
                "type": "opennebula"
            },
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "flavour": "medium",
                "hostname": "node-opennebula-medium-5.example.net",
                "ipaddr": "192.168.2.15",
                "type": "opennebula"
            }
        ],
        "retired_at": null,
        "tenant": {
            "active": true,
            "created_at": "2021-12-21T11:18:22.036233+00:00",
            "id": 2,
            "is_admin": false,
            "name": "tenant",
            "retired_at": null,
            "ssh_key": "Boo!"
        }
    }
}
```
- Request it again to verify that Duffy refuses to hand out more nodes than are available:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> http --json post http://localhost:8080/api/v1/sessions tenant_id=2 nodes_specs:='[{"quantity": 1, "type": "physical", "distro_type": "fedora", "distro_version": "35"}, {"quantity": 2, "type": "virtual", "flavour": "medium", "distro_type": "fedora", "distro_version": "35"}]'
HTTP/1.1 422 Unprocessable Entity
content-length: 122
content-type: application/json
date: Tue, 21 Dec 2021 12:23:16 GMT
server: uvicorn

{
    "detail": "can't reserve nodes: type=<NodeType.physical: 'physical'> quantity=1 distro_type='fedora' distro_version='35'"
}
```
- Release the reserved nodes by setting the `active` flag to `false` (check `active` and `retired_at` in the response):
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--session-workflow)> http --json put http://localhost:8080/api/v1/sessions/1 active:=false
HTTP/1.1 200 OK
content-length: 799
content-type: application/json
date: Tue, 21 Dec 2021 12:25:33 GMT
server: uvicorn

{
    "action": "put",
    "session": {
        "active": false,
        "created_at": "2021-12-21T11:22:13.362065+00:00",
        "id": 1,
        "nodes": [
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "hostname": "node-seamicro-2.example.net",
                "ipaddr": "192.168.0.12",
                "type": "seamicro"
            },
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "flavour": "medium",
                "hostname": "node-opennebula-medium-3.example.net",
                "ipaddr": "192.168.2.13",
                "type": "opennebula"
            },
            {
                "comment": null,
                "distro_type": "fedora",
                "distro_version": "35",
                "flavour": "medium",
                "hostname": "node-opennebula-medium-5.example.net",
                "ipaddr": "192.168.2.15",
                "type": "opennebula"
            }
        ],
        "retired_at": "2021-12-21T12:25:33.928737+00:00",
        "tenant": {
            "active": true,
            "created_at": "2021-12-21T11:18:22.036233+00:00",
            "id": 2,
            "is_admin": false,
            "name": "tenant",
            "retired_at": null,
            "ssh_key": "Boo!"
        }
    }
}
```
